subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps.c

```diff
@@ -0,0 +1,1585 @@
+	}
+	index.expected_count = request->prediction_count;
+
+	ret = cJSON_AddNumberToObject(data_obj, PGPS_JSON_PRED_COUNT,
+				      request->prediction_count);
+	if (ret == NULL) {
+		LOG_ERR("Failed to add pred count to P-GPS request:%d", err);
+		err = -ENOMEM;
+		goto cleanup;
+	}
+	ret = cJSON_AddNumberToObject(data_obj, PGPS_JSON_PRED_INT_MIN,
+				      request->prediction_period_min);
+	if (ret == NULL) {
+		LOG_ERR("Failed to add pred int min to P-GPS request:%d", err);
+		err = -ENOMEM;
+		goto cleanup;
+	}
+	ret = cJSON_AddNumberToObject(data_obj, PGPS_JSON_GPS_DAY,
+				      request->gps_day);
+	if (ret == NULL) {
+		LOG_ERR("Failed to add gps day to P-GPS request:%d", err);
+		err = -ENOMEM;
+		goto cleanup;
+	}
+	ret = cJSON_AddNumberToObject(data_obj, PGPS_JSON_GPS_TIME,
+				      request->gps_time_of_day);
+	if (ret == NULL) {
+		LOG_ERR("Failed to add gps time to P-GPS request:%d", err);
+		err = -ENOMEM;
+		goto cleanup;
+	}
+
+	err = json_send_to_cloud(pgps_req_obj);
+	if (!err) {
+		state = PGPS_REQUESTING;
+	}
+
+cleanup:
+	cJSON_Delete(pgps_req_obj);
+
+	return err;
+}
+
+int nrf_cloud_pgps_request_all(void)
+{
+	uint16_t gps_day;
+	uint32_t gps_time_of_day;
+	int err;
+
+	if (nrf_cloud_pgps_loading()) {
+		return 0;
+	}
+
+	npgps_reset_block_pool();
+
+	index.stale_server_data = false;
+	err = npgps_get_time(NULL, &gps_day, &gps_time_of_day);
+	if (err) {
+		/* unknown time; ask server for newest data */
+		gps_day = 0;
+		gps_time_of_day = 0;
+	}
+
+	struct gps_pgps_request request = {
+		.prediction_count = NUM_PREDICTIONS,
+		.prediction_period_min = PREDICTION_PERIOD,
+		.gps_day = gps_day,
+		.gps_time_of_day = gps_time_of_day
+	};
+
+	return nrf_cloud_pgps_request(&request);
+}
+
+int nrf_cloud_pgps_preemptive_updates(void)
+{
+	/* keep unexpired, read newer subsequent to last */
+	struct gps_pgps_request request;
+	int current = index.cur_pnum;
+	int valid_remaining = index.header.prediction_count - (current + 1);
+	uint16_t period_min = index.header.prediction_period_min;
+	uint16_t gps_day = 0;
+	uint32_t gps_time_of_day = 0;
+
+	if (current == 0xff) {
+		return nrf_cloud_pgps_request_all();
+	}
+	if (valid_remaining >= REPLACEMENT_THRESHOLD) {
+		LOG_DBG("Updates not needed yet");
+		return 0;
+	}
+
+	if (handler) {
+		handler(PGPS_EVT_LOADING, NULL);
+	}
+
+	LOG_INF("Replacing %d oldest predictions", current);
+	discard_oldest_predictions(current);
+	npgps_gps_sec_to_day_time(index.end_sec, &gps_day, &gps_time_of_day);
+
+	request.gps_day = gps_day;
+	request.gps_time_of_day = gps_time_of_day;
+	request.prediction_count = current;
+	request.prediction_period_min = period_min;
+	return nrf_cloud_pgps_request(&request);
+}
+
+int nrf_cloud_pgps_inject(struct nrf_cloud_pgps_prediction *p,
+			  const struct gps_agps_request *request,
+			  const int *socket)
+{
+	int err;
+	int ret = 0;
+	struct gps_agps_request remainder;
+	struct gps_agps_request processed;
+
+	memcpy(&remainder, request, sizeof(remainder));
```

Comment:
`request` could be NULL-checked first

Reply:
I'd already changed it so that, if set to NULL, it is assumed only ephemerides are needed.